src/mpi/datatype/typerep/src/typerep_util.h

```diff
@@ -128,24 +128,25 @@ static inline void BASIC_convert128(const char *src, char *dest)
 }
 
 #if (BLENDIAN == 1)
-#define BASIC_convert(src, dest)               \
-{                                           \
-    register int type_byte_size = sizeof(src); \
-    switch(type_byte_size)                     \
-    {                                          \
-        case 1:                                \
-            dest = src;                        \
-            break;                             \
-        case 2:                                \
-            BASIC_convert16(src, dest);        \
-            break;                             \
-        case 4:                                \
-            BASIC_convert32(src, dest);        \
-            break;                             \
-        case 8:                                \
-            BASIC_convert64(src, dest);        \
-            break;                             \
-    }                                          \
+/* Note: use `uint` to supress warnings on left-shift */
+static inline void BASIC_convert(const void *src, void *dest, int size)
```

Comment:
This implementation do not support unaligned buffers.

You could define the following helper macro:
```c
#define BASIC_convert_bits(src, dest, bits) \
do {                                        \
    uint##bits##_t s, d;                    \
    memcpy(&s, src, sizeof(s));             \
    BASIC_convert##bits(s, d);              \
    memcpy(dest, &d, sizeof(d));            \
} while (0)
```

and then use it like this:
```c
{
    switch (size) {
        case 1:
            BASIC_convert_bits(src, dest,  8);
            break;
        case 2:
            BASIC_convert_bits(src, dest, 16);
            break;
        case 4:
            BASIC_convert_bits(src, dest, 32);
            break;
        case 8:
            BASIC_convert_bits(src, dest, 64);
            break;
        default:
            MPIR_Assert(0);
    }
}
```

Reply:
> This implementation do not support unaligned buffers.

I know, but internally we are debating the balance of usefulness and performance. Only when derived types involves `struct` or `hindexed` types will bring alignment issue. The current yaksa code doesn't support unpack_external on struct. And the solution is probably to build a separate slow path for `struct` types. So we'll solve the alignment issues then and there. As for `hindexed` type, a type that breaks the alignment is not useful -- since the user can't really portabally or performantly use the data on such type, so we don't expect the issue to have any real impact. At lease we are reluctant to bring down the performance and add complexity for such impractical cases (until someone convince us otherwise).

Reply:
Oh, of course, you have to define the following trivial macro
```
#define BASIC_convert8(src, dest) \
{                                 \
  dest = src;                     \
}
```

Reply:
> Only when derived types involve `struct` or `hindexed` types will bring alignment issue

Mmm... I'm not following you. IIRC, the MPI standard says that user buffers may be unaligned, and the MPI implementation has to cope with that hand handle the mess gracefully.

OK, perhaps you should forget about alignment for now. Later, you can use a fast code path if the buffer addresses are aligned, and go through the unaligned path with memcpy calls otherwise.

Reply:
> IIRC, the MPI standard says that user buffers may be unaligned

I'm not sure about this claim. Pehaps this was from an old version of the standard.

Reply:
> OK, perhaps you should forget about alignment for now. Later, you can use a fast code path if the buffer addresses are aligned, and go through the unaligned path with memcpy calls otherwise.

Yeah, the priority would be to take care of the `struct` first, which I don't have a simple solution (in yaksa) yet.